products/data_warehouse/backend/hogql_fixer_ai.py

```diff
@@ -0,0 +1,174 @@
+Fix the errors in the HogQL query below and only return the new updated query in your response.
+
+- Don't update any other part of the query if it's not relevant to the error, including rewriting shorthand clickhouse SQL to the full version.
+- Don't change the capitalisation of the query if it's not relevant to the error, such as rewriting `select` as `SELECT` or `from` as `FROM`
+- There may also be more than one error in the syntax.
+
+Below is the current HogQL query and the error message
+"""
+
+
+class HogQLQueryFixerTool(MaxTool):
+    name: str = "fix_hogql_query"
+    description: str = "Fixes any error in the current HogQL query"
+    thinking_message: str = "Fixing errors in the SQL query"
+    root_system_prompt_template: str = SQL_ASSISTANT_ROOT_SYSTEM_PROMPT
+
+    def _run_impl(self) -> tuple[str, str | None]:
+        database = create_hogql_database(self._team_id)
+        hogql_context = HogQLContext(team_id=self._team_id, enable_select_queries=True, database=database)
+
+        serialized_database = serialize_database(hogql_context)
+        schema_description = "\n\n".join(
+            (
+                f"Table `{table_name}` with fields:\n"
+                + "\n".join(f"- {field.name} ({field.type})" for field in table.fields.values())
+                for table_name, table in serialized_database.items()
+                # Only the most important core tables, plus all warehouse tables
+                if table_name in ["events", "groups", "persons"]
+                or table_name in database.get_warehouse_tables()
+                or table_name in database.get_views()
+            )
+        )
+
+        prompt = ChatPromptTemplate.from_messages(
+            [
+                (
+                    "system",
+                    SYSTEM_PROMPT.replace("{{schema_description}}", schema_description),
+                ),
+                (
+                    "user",
+                    USER_PROMPT
+                    + "\n\n<hogql_query>"
+                    + "{{{hogql_query}}}"
+                    + "</hogql_query>"
+                    + "\n\n<error>"
+                    + "{{{error_message}}}"
+                    + "</error>",
+                ),
+            ],
+            template_format="mustache",
+        )
+
+        for i in range(3):
+            try:
+                chain = prompt | self._model
+                result = chain.invoke(self.context)
+                parsed_result = self._parse_output(result, hogql_context)
+                break
+            except PydanticOutputParserException as e:
+                prompt += f"\n\nWe've ran this prompt {i+1} time{'s' if i+1 > 1 else ''} now. The newly updated query gave us this error: {e.validation_message}"
+        else:
+            return "", None
+
+        return parsed_result, parsed_result
+
+    @property
+    def _model(self):
+        return ChatOpenAI(model="gpt-4o", temperature=0, disable_streaming=True).with_structured_output(
```

Comment:
nit: I think this is the pattern we have been following elsewhere but feels like things like the model should be defined in a constant somewhere as I imagine we'll need to update it across the app within the next few months

Reply:
Yep - good nit